### ISSUE TITLE
This library vs paginateScan

### ISSUE THREAD

**Issue by paul-uz** [shelfio/dynamodb-parallel-scan#224]
I seem to be having an issue with paginateScan in the v3 SDK being slow to retrieve 5000 items from a 16mb table. It takes about 7 seconds on cold boot, with total segments equalling being set to how many mb the table is and a pageSize of 200 (I'm still not understanding the role of pageSize properly)

Do you think this library will perform better?

**Comment by vladgolubev** [comment]
Yes, `paginateScan` is sequential. This library works by making requests in parallel.

You can check the source code of aws-sdk to see that `paginateScan` is making 1 request at a time
![2023-02-23T13 50 31-PJpYZBRE](https://user-images.githubusercontent.com/3817380/220911430-8f3a88a5-dd1b-4d6b-8232-b7d33ee06783.png)

**Comment by paul-uz** [comment]
I did implement a parallel version of it, by running multiple in a promise all, but I suspect my code isn't the most efficient. 

Will definitely try this library out soon

**Comment by paul-uz** [comment]
So I have replaced my scan with this library, and i'm still getting seemingly slow speeds.

At ~9500 records now, with a table size of 39mb, it takes 10s to retrieve on cold boot. Does that seem right? Looks awfully slow to me :/

**Comment by vladgolubev** [comment]
https://medium.com/shelf-io-engineering/how-to-scan-a-23-gb-dynamodb-table-in-1-minute-110730879e2b
<img width="818" alt="2023-02-23T18 28 57-acpNtLWr" src="https://user-images.githubusercontent.com/3817380/220984087-a109db68-94d5-4ef3-a497-f568ecb525b6.png">


Is this might be a problem? Check our hash key uniqueness

**Comment by paul-uz** [comment]
Our hash keys (partition keys) are all UUIDs, so each one is unique.

**Comment by paul-uz** [comment]
@vladgolubev I notcied in your article, you show this code

```javascript
for await (const items of stream) {
    console.log(items); // 10k items here
    // do some async processing to offload data from memory
    // and move on to cosuming the next chunk
    // scanning will be paused for that time
  }
```

Could you explain more about the comments here about offloading the data? I'm wondering if this could help me.

**Comment by vladgolubev** [comment]
Scanning might be slow if you take 1 page of data which takes ~1s, and then spend 10s to process scanned data. It's faster if you scan all data at once, to collect into 1 array, and then spend time on more costly data processing.

**Comment by paul-uz** [comment]
Hmm, I'm not doing much in terms of processing the data, other than sorting and splicing the array of all items.

**Comment by paul-uz** [comment]
I run it in Lambda, with 1024mb of memory, but have been experimenting with 2048mb as well

**Comment by vladgolubev** [comment]
Is your table using ON_DEMAND or PROVISIONED capacity mode?

**Comment by paul-uz** [comment]
On Demand

**Comment by vladgolubev** [comment]
At this point, I think AWS Support will be better to assist. I ran out of ideas ¯\_(ツ)_/¯

**Comment by bilegjargal-jargalsaikhan** [comment]
Hey 👋 , not related to this topic but how could we pass credentials to the `parallelScanAsStream` function? Any examples?
Thanks :)

**Comment by harazdovskiy** [comment]
@bilegjargal-jargalsaikhan try new version new [v3.5.3](https://github.com/shelfio/dynamodb-parallel-scan/releases/tag/v3.5.3)
<img width="1281" alt="Screenshot 2023-07-12 at 22 35 08" src="https://github.com/shelfio/dynamodb-parallel-scan/assets/19535641/08be9cd3-754f-4aff-bfb8-fe0f799b67dc">
